### ISSUE TITLE
snake_case checking doesn't work with characters outside a-z0-9 range

### ISSUE THREAD

**Issue by msaari** [WordPress/WordPress-Coding-Standards#1891]
## Bug Description

I get `Variable "$lähtöaika" is not in valid snake_case format, try "$lähtöaika"` error, when clearly the variable name is in valid snake_case format.

## Minimal Code Snippet
```php
$lähtöaika = true;
```

This is because the `isSnakeCase()` function only allows characters within `[a-z0-9_]`. However, this is not the whole set of allowed characters in PHP variable names. Quoting from the [PHP manual](https://www.php.net/manual/en/language.variables.basics.php):

> Variable names follow the same rules as other labels in PHP. A valid variable name starts with a letter or underscore, followed by any number of letters, numbers, or underscores. As a regular expression, it would be expressed thus: `^[a-zA-Z_\x80-\xff][a-zA-Z0-9_\x80-\xff]*$`

If the `isSnakeCase()` function is changed to use this regular expression (minus the uppercase A-Z), the problem is gone:

```php
public static function isSnakeCase( $var_name ) {
	return (bool) preg_match( '/^[a-z_\x80-\xff][a-z0-9_\x80-\xff]*$/', $var_name );
}
```

This is not perfect, as the 128-255 character range contains uppercase letters, so some non-snake variable names would still be accepted: `$lähtÖaika` would not trigger an error. I'm not sure if checking for these cases would be worth making the regex a lot more complicated, though.

## Error Code
`WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase`

## Environment

| Question               | Answer
| ------------------------| -------
| PHP version             | 7.3.11
| PHP_CodeSniffer version | 3.4.2
| WPCS version            | 2.2.1
| WPCS install type       | global, VS Code plugin
| IDE (if relevant)       | VS Code 1.44.2

## Tested Against `develop` branch?
- [X] I have verified the issue still exists in the `develop` branch of WPCS.

**Comment by msaari** [comment]
If this is restricted on purpose to encourage variable names in English – which is fine, I suppose – getting the snake_case complaint is wrong and there should be a separate sniff for having non `[a-z0-9_]` letters in variable names.

**Comment by dingo-d** [comment]
Did you try whitelisting the variable name?

https://github.com/WordPress/WordPress-Coding-Standards/wiki/Customizable-sniff-properties#mixed-case-property-name-exceptions

**Comment by msaari** [comment]
Why? Whitelisting the variable name doesn't solve the general problem of WPCS complaining about snake_ case when a variable has non-[a-z_0-9] characters.

Whitelisting the variable name would be an acceptable solution if I wanted to use a non-snake_case variable and avoid the error, but here the variable name is in valid snake_case format, so WPCS should not complain about snake_case.

**Comment by dingo-d** [comment]
You could extend the sniff in your own ruleset and replace the regex (the `isSnakeCase` is a static public method).

The WPCS is English based, as code is usually written in English, I guess that's the reason for the regex being that way.

@jrfnl what's your thought about this?

**Comment by msaari** [comment]
I understand this is because the code is written in English and I understand WPCS doesn't need to really support non-English variable names (and they may be a bad idea in the first place; I don't usually do it myself either, just happened to notice this in a hobby project of mine), but the error message is still wrong and misleading.

If WPCS wants to enforce no non-[a-z0-9] characters in variable that should be a separate sniff, and not baked into snake_case sniff as this has nothing to do with snake_case. This would probably make a reasonable sniff of its own.

**Comment by jrfnl** [comment]
I agree with @msaari that this is a bug. I have some code already set up to fix this, but am not completely happy with it yet. Once I am, a set of methods to check often-used naming conventions will go into PHPCSUtils and WPCS will defer to those instead of using the current WPCS native check.

I can't promise this will go into WPCS 3.0.0, but am hoping it will be ready in time. If not, you can expect it in a patch/minor release in the 3.x series soon after.

--- 

For those interesting in the technical details:
The problem with checking text strings like this is, of course, character encoding, and by extension file encoding.
WPCS presumes that files are encoded in UTF-8 and checking this reliably with a unicode based regex is perfectly doable (and for that, the code is ready).
But PHPCS allows for setting a different file encoding using the `--encoding` parameter. If the encoding is changed, we have a whole different problem as then we'd need to convert text between encodings to be able to check the text. Converting relies on PHP extensions which aren't always available, so becomes unreliable based on that alone.
Aside from that, I haven't figured out a way to generate test case files in a non-standard encoding, so for now, haven't been able to test whether the potential solutions I have come up could work.

Either way, once I figure it out, we should have a fix for this. (oh and if anyone can generate some test case files based on my specs in, for instance, the Big-5 encoding, that would be very helpful!)
